### ISSUE TITLE
Xamarin support

### ISSUE THREAD

**Issue by Project4MobileDev** [stryker-mutator/stryker-net#465]
I have a Xamarin project where I'm running stryker and it's complaining that Xamarin.Forms.Xaml assembly failed to resolve. This then seems to cause a problem where it tries to mutate the InitializeComponents method which is part of Xamarin which it fails to do. To solve this I tried excluding files in the /views/ folder so it wouldn't try to mutate any of the .xaml.cs files. It says they're being excluded but then it appears that it's still trying to mutate them anyway. I'll copy my config file and log file below. Does stryker normally work with Xamarin?

**Environment**
Windows 10
.NET Command Line Tools (2.2.203)

**Config file**
```
{
  "stryker-config": {
    "files-to-exclude": [
      "/App.xaml.cs",
      "/Views/"
    ]    
  }
}
```

[log-20190419.txt](https://github.com/stryker-mutator/stryker-net/files/3098677/log-20190419.txt)

**Comment by richardwerkman** [comment]
Hi @Project4MobileDev thanks for reporting this bug. We've seen this before in another issue. I've looked at this before and couldn't find why Stryker is trying to mutate the excluded files. I'll have another go at this :+1:

**Comment by richardwerkman** [comment]
To answer your question about xamarin support, we do not officially support xamarin. However we haven't tested if it runs on xamarin projects. 

After looking into your logfile I can say with certainty that there are no mutations being made in your excluded file. I'll try to explain what's going on, for as far as I do understand from your log output.

We build your project with `dotnet build` to see if your project is buildable and your project builds correctly. Then stryker mutates your project and tries to compile again. However the second time we compile with the roslyn compiler and not with `dotnet build`. It looks like the roslyn compiler has trouble with compiling the `App.xaml.cs` file. 

When roslyn returns us compile errors we assume they are caused by our mutations so we rollback the mutation that caused the error. That is why you see the log statement saying your mutations in `App.xaml.cs` will be rollbacked. There are no mutations however so after rollbacking we still cannot compile the file and stryker crashes...

This might be an interesting case to see if we can get stryker to support xamarin projects. If `dotnet build` can compile the project then we should be able to do too with roslyn. We just have to figure out what info to pass to roslyn.

I'll rename the issue to "Xamarin support". So we can investigate this further under this issue.

**Comment by richardwerkman** [comment]
My first thought is that we might not have to feed the syntax trees of any `*.xaml.cs` file as it does not contain pure c#. If we only feed the true c# files to roslyn it might produce a valid assembly that we can run stryker on. I do not have a xamarin to test this on at the moment.

**Comment by Project4MobileDev** [comment]
Ah cool, thanks for the explanation. Yeah adding Xamarin support would be really useful thanks. What's the best way to progress this? Do you want me to attach a sample Xamarin project?

**Comment by Mobrockers** [comment]
If you could provide us with a small, known working xamarin project, that would allow us to investigate.

**Comment by Project4MobileDev** [comment]
Here's a link to a simple Xamarin solution, it's just a blank out-of-the-box Prism solution, it's the same thing you'd be able to get by just creating a new Xamarin/Prism solution in Visual Studio. Any Xamarin solution should be able to reproduce the bug but this is the one I used. I can't attach the zip file directly because it's too big so I've had to use a link, the link will expire in one week but hopefully that should be fine. Let me know if you need anything else. Cheers

https://we.tl/t-LNP5iPWABf

**Comment by Project4MobileDev** [comment]
Did you get the project okay? The link expires in two days. What do you think the time scale on adding Xamarin support might be roughly? Thanks again

**Comment by richardwerkman** [comment]
I've downloaded your project and will try to run stryker on it soon. After I've played around with it I might be able to estimate what time it would take. This fully depends on how many changes we have to make in stryker to compile the assembly and run the tests agains the mutated assembly. 

For me xamarin support doesn't have the highest priority. I think multi project runs and some bugs are more important right now.

If you want to use stryker soon you could try to start on this feature yourself? I could help you by giving pointers and code with you :)

**Comment by richardwerkman** [comment]
@Project4MobileDev I've succesfully run stryker on your example xamarin project. I'll create a PR right now.

Could you verify that the PR code can run on your real xamarin project as well?

**Comment by Project4MobileDev** [comment]
Ah awesome, thanks for looking at that. Yeah sure, could you let me know how I would go about running stryker from your PR branch?

**Comment by richardwerkman** [comment]
No problem, it turned out to be no big deal :). Thanks for helping out with some testing! The steps to run stryker from code on a project are explained in our [contribution guide](https://github.com/stryker-mutator/stryker-net/blob/master/CONTRIBUTING.md#example-steps-to-run-stryker-on-stryker). 

Of course you'll have to checkout the branch `xamarin-support` instead of default `master`.
If you need more help after reading the contribution guide let me know!

**Comment by Project4MobileDev** [comment]
Alright I've followed the steps in the contribution guide but I'm hitting a problem for some reason:

"Stryker.NET failed to mutate your project. For more information see the logs below:
No .csproj file found, please check your project directory at C:\R\stryker-net\src\Stryker.CLI\Stryker.CLI\bin\Debug\netcoreapp2.1"

I've set the working directory of the debug profile to the directory containing the .csproj file for the unit tests:
C:\R\App\tests\UnitTests

Any ideas what I might be doing wrong?

**Comment by richardwerkman** [comment]
Yeah, I think you created the debug profile but did not select it to run. From your output I can see it runs from the default directory inside stryker.

You can select a debug profile next to the run button in visual studio :+1:

**Comment by Project4MobileDev** [comment]
Ah yeah that was it, I didn't have the debug profile selected. Yeah it runs successfully on our Xamarin project, very cool :)

**Comment by richardwerkman** [comment]
Awesome to hear! 🎉 I guess the PR code will be released with the next minor release 0.11.0. Have fun using stryker!
